rcon/commands.py

```diff
@@ -73,78 +89,119 @@ class ServerCtl:
     set password not implemented on purpose
     """
 
-    def __init__(self, config, auto_retry=1):
+    def __init__(self, config, auto_retry=1, pool_size=20):
         # .env fed config from rcon.SERVER_INFO
         self.config = config
-        self.conn = None
-        # self._connect()
         self.auto_retry = auto_retry
+        self.mu = threading.RLock()
+        self.idles: list[HLLConnection] = []
+        self.numOpen = 0
+        self.maxOpen = 20
+        self.maxIdle = pool_size
+
+    @contextmanager
+    def with_connection(self) -> HLLConnection:
+        logger.debug("Waiting to acquire lock")
+        with self.mu:
+            if len(self.idles) != 0:
+                logger.debug("acquiring connection from idle pool")
+                conn = self.idles.pop()
+
+            elif self.numOpen >= self.maxOpen:
+                logger.debug("Max connections already open, waiting for connection returned to pool")
```

Comment:
so the pool size should always be equal or higher than the number of threads used by gunicor so this doesn't happen

Reply:
Not quite, I mean, there might be _some_ overhead on certain actions that require the RCon actions to be performed. In this case, the rcon connection would be returned to the pool much sooner than the gunicorn process is done.

One thing were this would not be true would be the team view: It will utilize as much connections as possible (up to the number of threads in the RecordedRcon ThreadPool), while it is still just one gunicorn process. In this case, the pool _might_ hit the max open connections limit and new actions that wait for a connection will be kept on hold once a connection is available.

Maybe we can reduce the wait time from 1 second to some lower value. But I assume that people will not see this very often based on the default pool size as well as the caching going around for most of the data.